tests/unit/test_logging.py

```diff
@@ -0,0 +1,244 @@
+    record_compiler.handle(record)
+    for key in KEYS_IN_JSON_LOG:
+        assert key in record.__dict__
+
+
+def test_json_formatter(caplog):
+    """Test that the JsonFormatter works like expected"""
+    formatter = JsonFormatter()
+
+    assert not caplog.records
+    log = logging.getLogger("test_json_formatter")
+    log.setLevel("INFO")
+    log.error("This is a %s", "test")  # test with arg to make sure this isn't lost
+
+    assert len(caplog.records) == 1
+    record = caplog.records[0]
+
+    # we get an error if we skip the RecordCompiler
+    with pytest.raises(KeyError):
+        formatter.format(record)
+
+    # use the record compiler to stick in the extra fields
+    record_compiler = RecordCompiler(config=DEFAULT_CONFIG)
+    record_compiler.handle(record)
+
+    # format with the json formatter
+    output = formatter.format(record)
+
+    # load back into json format and verify the added keys are there
+    json_log = json.loads(output)
+    for key in ADDED_KEYS:
+        assert key in json_log
+
+    # make sure the message is what we expected
+    assert json_log["message"] == "This is a test"
+
+
+@pytest.mark.parametrize(
+    "log_format, formatter_class",
+    [
+        (None, JsonFormatter),
+        ("%(message)s", logging.Formatter),
+    ],
+)
+def test_formatter_selection(
+    log_format: Optional[str], formatter_class: type[logging.Formatter]
+):
+    """Make sure the proper formatter is selected based on the config."""
+    config = LoggingConfig(
+        service_name=VALID_SERVICE_NAME,
+        service_instance_id=VALID_INSTANCE_ID,
+        log_format=log_format,
+    )
+    log = logging.getLogger("test_formatter_selection")
+    configure_logging(config=config, logger=log)
+
+    handlers = log.handlers
+    assert isinstance(handlers[0].formatter, formatter_class)
+
+
+def test_reconfiguration_of_existing_loggers():
+    """Ensure applying configuration to an existing logger works as expected."""
+    log = logging.getLogger("reconfig")
+    trace = 5
+
+    assert len(log.handlers) == 0
+
+    config = LoggingConfig(
+        log_level="TRACE",  # can verify as well that the custom level exists
+        service_name=VALID_SERVICE_NAME,
+        service_instance_id=VALID_INSTANCE_ID,
+        log_format="%(timestamp)s - %(msg)s",
+    )
+
+    configure_logging(config=config, logger=log)
+    assert log.getEffectiveLevel() == trace
+    assert isinstance(log.handlers[0].formatter, logging.Formatter)
+
+
+def test_root_config():
+    """Test that the root logger is configured by default."""
+    root = logging.getLogger()
+    level = root.level
+
+    for handler in root.handlers:
+        assert not isinstance(handler, RecordCompiler)
+
+    configure_logging(config=DEFAULT_CONFIG)
+
+    assert any([isinstance(handler, RecordCompiler) for handler in root.handlers])
+
+    for handler in root.handlers:
+        if isinstance(handler, RecordCompiler):
+            root.removeHandler(handler)
```

Comment:
Might still be a problem if the test fails before cleanup. This should fix or use a fixture:
```suggestion
    try:
        configure_logging(config=DEFAULT_CONFIG)
        root_handlers = list(root.handlers).copy()
    finally:
        for handler in root.handlers:
            if isinstance(handler, RecordCompiler):
                root.removeHandler(handler)

    assert any([isinstance(handler, RecordCompiler) for handler in root_handlers])
```

Reply:
I went with a fixture and put it in the `utils` module so it can be used by other tests if needed. [5194b2f](https://github.com/ghga-de/hexkit/pull/77/commits/5194b2f77a0239dd88ff4432d4f6161f7cd89a71)